Add the splinter upgrade command

leebradley opened the pull request:
Add "splinter upgrade" command that imports data from a yaml state store
to a database.

agunde406 commented on cli/src/action/database/mod.rs:
```diff
@@ -63,6 +77,227 @@ impl Action for MigrateAction {
+            CliError::ActionError("'state_dir' is not a valid UTF-8 string".to_string())
+        }
+
+        let circuits_location = Path::new(&state_dir).join("circuits.yaml");
```
The state files can be added as const above.

leebradley replied:
Done!

agunde406 commented on cli/src/action/database/mod.rs:
```diff
@@ -63,6 +77,227 @@ impl Action for MigrateAction {
+
+        Self::import_command(&mut io::stdout(), &state_dir, || {
+            // Verify database connection before setting up pool
+            PgConnection::establish(&database_uri[..]).map_err(|err| {
```
Commit message says that only Postgress is supported, however this command should support sqlite as well.

agunde406 commented on cli/src/main.rs:
```diff
@@ -948,7 +948,25 @@ fn run<I: IntoIterator<Item = T>, T: Into<OsString> + Clone>(args: I) -> Result<
+            SubCommand::with_name("upgrade")
+                .about("Upgrade splinter")
+                .arg(
+                    Arg::with_name("state_dir")
```
Please add `.value_name("state-dir")`

leebradley replied:
Done

agunde406 commented on cli/src/action/database/mod.rs:
```diff
@@ -23,12 +23,26 @@ use diesel::{connection::Connection as _, pg::PgConnection};
+use splinter::admin::store::error::AdminServiceStoreError;
+use splinter::admin::store::yaml::YamlAdminServiceStore;
+use splinter::admin::store::AdminServiceStore;
+use std::env;
```
Please rearrange your use statements into blocks in the order std, external crate, internal crate. We try to standardize this format for the use statements.

peterschwarz replied:
With a new line in between, so the formatter doesn't resort them.

This is one thing we can't (currently) catch with the linter, but is a team preference.

leebradley replied:
Fixed!

leebradley commented:
Made changes in response to feedback.

Major changes:
- Added experimental feature `upgrade`
   - To facilitate this, I moved the importer to its own module under `cli::action::database::upgrade`
- Split out commit for `SplinterEnvironment`
- Added support for sqlite

Other changes:
- Replaced instances of `write!` and `writeln!` with `info!`
  - Because there's no easy way to capture or validate the output of the `info!` macro, I modified the unit test to validate using database output instead of the command output
- Added value-name of "state-dir" for state argument
- Added more information to argument help
- Moved some static values to consts

I also found that because `YamlAdminServiceStore` creates invalid circuit objects (apparently they lack endpoints, and this is by design?), it corrupts the data in a way that makes reading those circuits impossible. I work around this by filling in those endpoints during the import process.

This is ready for another round of review.

leebradley commented on cli/Cargo.toml:
```diff
@@ -46,11 +46,12 @@ whoami = "0.7.0"
 
```
Fixed. Put everything behind an `update` feature

leebradley commented on cli/src/action/database/mod.rs:
```diff
@@ -63,6 +77,227 @@ impl Action for MigrateAction {
+        state_dir: &str,
+        get_db_store: impl FnOnce() -> Result<Box<dyn AdminServiceStore>, CliError>,
+    ) -> Result<(), CliError> {
+        writeln!(stdout, "Upgrading splinterd state")?;
```
Fixed. I'd initially done this for testing the console output. Unfortunately I can't really find a good way to capture the output using log macros `info!` and whatnot, but it's probably not a big deal.

leebradley commented on cli/src/action/database/sqlite.rs:
```diff
@@ -70,48 +67,20 @@ pub fn sqlite_migrations(connection_string: String) -> Result<(), CliError> {
-/// Otherwise, returns `/var/lib/splinter/splinter_state.db`
+/// Gets the splinter default state path, creating it if it does not exist. Creates a db file with
+/// the name splinter_state.db.
 pub fn get_default_database() -> Result<String, CliError> {
```
Done. I split this out into another commit.

leebradley commented on cli/src/main.rs:
```diff
@@ -948,7 +948,25 @@ fn run<I: IntoIterator<Item = T>, T: Into<OsString> + Clone>(args: I) -> Result<
+
+        app = app.subcommand(
+            SubCommand::with_name("upgrade")
+                .about("Upgrade splinter")
```
Updated to now say "Upgrade splinter by importing YAML state files to a database"

agunde406 commented on cli/src/main.rs:
```diff
@@ -951,6 +951,27 @@ fn run<I: IntoIterator<Item = T>, T: Into<OsString> + Clone>(args: I) -> Result<
+                .arg(
+                    Arg::with_name("state_dir")
+                        .value_name("state-dir")
+                        .short("S")
```
Can you also add `.long("state-dir")` here and `.long("connect")` below

leebradley replied:
Done

agunde406 commented on cli/src/action/database/upgrade.rs:
```diff
@@ -0,0 +1,493 @@
+
+impl Action for ImportFromYamlAction {
+    fn run<'a>(&mut self, arg_matches: Option<&ArgMatches<'a>>) -> Result<(), CliError> {
+        let (state_dir, database_uri) = match arg_matches {
```
If I provide a state-dir but not a database uri (connect value) I expect that database to be in the same state dir as what I provided. This does not happen now, instead it puts the database in the default location.

agunde406 replied:
It would also be useful to log the state directory being used and what database the state is being moved to for debugging purposes. (with absolute paths for files)

leebradley replied:
Fixed!
- Code now looks for the database in the state directory if sqlite feature is enabled
- `splinter upgrade` now shows what directory it's using for the state and what database uri it's using (with absolute paths)

agunde406 commented:
Also we need to write a man page for this command but that can go in another PR

agunde406 commented on cli/src/action/database/upgrade.rs:
```diff
@@ -0,0 +1,492 @@
+            .map_err(|e| CliError::ActionError(format!("could not move proposals file: {}", e)))?;
+
+        info!(
+            "Successfully imported {} node(s) and {} proposal(s)",
```
Should nodes be circuits (in the log and in the import results)?

leebradley replied:
Yep! Fixed
